pallets/utxo/src/sign.rs

```diff
@@ -0,0 +1,239 @@
+//!    See [Public] and [SignatureData].
+
+use crate::{Transaction, TransactionOutput};
+
+use chainscript::context::ParseResult;
+pub use chainscript::sighash::SigHash;
+use chainscript::sighash::{InputMode, OutputMode};
+use codec::{Decode, DecodeAll, Encode};
+use frame_support::sp_io::crypto;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
+use sp_core::{sr25519, H256};
+use sp_runtime::traits::{BlakeTwo256, Hash};
+use sp_std::prelude::*;
+use variant_count::VariantCount;
+
+/// Transaction data comitted to in a signature.
+#[derive(Eq, PartialEq, Clone, Encode)]
+pub struct TransactionSigMsg {
+    /// Sighash
+    sighash: SigHash,
+    /// Information about inputs
+    inputs: TransactionInputSigMsg,
+    /// Information about outputs
+    outputs: TransactionOutputSigMsg,
+    /// OP_CODESEPARATOR position (or 0xffffffff if none seen so far)
+    codesep_idx: u32,
+}
+
+/// Transaction input data comitted to in a signature.
+#[derive(Eq, PartialEq, Clone, Encode)]
+enum TransactionInputSigMsg {
+    /// Commit to all inputs
+    CommitWhoPays {
+        outpoints: H256,
+        spending: H256,
+        index: u64,
+    },
+    /// Commit to this input only
+    AnyoneCanPay { outpoint: H256, spending: H256 },
+}
+
+/// Transaction output data comitted to in a signature.
+#[derive(Eq, PartialEq, Clone, Encode)]
+enum TransactionOutputSigMsg {
+    /// Commit to all transaction outputs
+    #[codec(index = 0x01)]
+    All { outputs: H256 },
+
+    /// Don't commit to any outputs (i.e. outputs can be changed freely)
+    #[codec(index = 0x02)]
+    None,
+
+    /// Commit to just one output that corresponds to the current input
+    #[codec(index = 0x03)]
+    Single { output: H256 },
+}
+
+impl TransactionSigMsg {
+    /// Create a `TransactionSigMsg` from a transaction, spent outputs, current index
+    /// and other context information according to given sighash.
+    ///
+    /// The `sighash` parameter specifies which parts of the transaction `tx` are signed. A list of
+    /// UTXOs corresponding to transaction inputs being spent is also required and has to be passed
+    /// in the `spending` parameter. No validation is done to ensure the UTXOs really match the
+    /// inputs, it is resposnsibility of the caller to verify it. The input being sign is passed in
+    /// the `index` argument and the index of the last `OP_CODESEPARATOR` is in `codesep_idx`.
+    ///
+    /// TODO This could be improved by pre-calculating hash values instead of hashing transaction
+    /// parts every time. Currently, all the hashes are recalculated from the scratch.
+    pub fn construct<AcctId: Encode>(
+        sighash: SigHash,
+        tx: &Transaction<AcctId>,
+        spending: &[TransactionOutput<AcctId>],
+        index: u64,
+        codesep_idx: u32,
+    ) -> Self {
+        let idx = index as usize;
+        assert!(spending.len() == tx.inputs.len());
+        assert!(idx < tx.inputs.len());
```

Comment:
Are these intentionally `assert` instead of `ensure`?

Reply:
Yes, it is the responsibility of the caller to ensure the index refers to a valid input and that there is exactly one spent UTXO per input.